pkg/services/ngalert/tests/fakes/rules.go

```diff
@@ -0,0 +1,341 @@
+package fakes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	models2 "github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+// FakeRuleStore mocks the RuleStore of the scheduler.
+type FakeRuleStore struct {
```

Comment:
Maybe for a follow-up PR, but is it worth just using the store.DBStore with sqlite3 when tests are complex enough to need this type? There's enough logic in here that it's an independent in-memory implementation rather than a deterministic fake, so maybe these complex tests are better expressed as integration tests between the DBStore and the calling package.

Reply:
This is what we did for Provisioning, it's definitely nicer. We didn't want to add further dependencies on this thing. It's a pain to maintain. Some other places use a mock instead of a fake, that's also been a huge improvement. I agree that this fake has gotten out of hand, and I've pared it down where I can.

I'm trying to limit the scope of this refactor a bit, it feels we're falling into the syndrome of finding something new that's broken every time we go to fix something. That's why I didn't spend too much time pulling apart particularly tough blobs like this. We removed a number of consumers of this fake already as part of this effort, that's significant progress.

The purpose of refactor is to unblock the state history work, and with this PR it seems we've reached that point. I'd like to wrap up any loose ends, and start filing remaining issues into the backlog rather than trying to fix the world here. I'll create an issue noting this, but I'm not sure if I'll be able to come back and actually rework these tests soon.